cdap-app-fabric/src/main/java/io/cdap/cdap/internal/capability/CapabilityApplier.java

```diff
@@ -0,0 +1,402 @@
+import io.cdap.cdap.common.ApplicationNotFoundException;
+import io.cdap.cdap.common.ArtifactNotFoundException;
+import io.cdap.cdap.common.InvalidArtifactException;
+import io.cdap.cdap.common.conf.CConfiguration;
+import io.cdap.cdap.common.namespace.NamespaceAdmin;
+import io.cdap.cdap.common.service.Retries;
+import io.cdap.cdap.common.service.RetryStrategies;
+import io.cdap.cdap.internal.app.deploy.ProgramTerminator;
+import io.cdap.cdap.internal.app.runtime.BasicArguments;
+import io.cdap.cdap.internal.app.services.ApplicationLifecycleService;
+import io.cdap.cdap.internal.app.services.ProgramLifecycleService;
+import io.cdap.cdap.internal.app.services.SystemProgramManagementService;
+import io.cdap.cdap.internal.entity.EntityResult;
+import io.cdap.cdap.proto.NamespaceMeta;
+import io.cdap.cdap.proto.ProgramType;
+import io.cdap.cdap.proto.id.ApplicationId;
+import io.cdap.cdap.proto.id.NamespaceId;
+import io.cdap.cdap.proto.id.ProgramId;
+import io.cdap.cdap.proto.metadata.MetadataSearchResponse;
+import io.cdap.cdap.proto.metadata.MetadataSearchResultRecord;
+import io.cdap.cdap.security.spi.authorization.UnauthorizedException;
+import io.cdap.cdap.spi.metadata.SearchRequest;
+import org.apache.twill.discovery.DiscoveryServiceClient;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import javax.annotation.Nullable;
+
+/**
+ * Class that applies capabilities
+ */
+public class CapabilityApplier {
+
+  private static final Logger LOG = LoggerFactory.getLogger(CapabilityApplier.class);
+  private static final Gson GSON = new Gson();
+  private static final int RETRY_LIMIT = 5;
+  private static final int RETRY_DELAY = 5;
+  private static final String CAPABILITY = "capability:%s";
+  private static final String APPLICATION = "application";
+  private static final ProgramTerminator NOOP_PROGRAM_TERMINATOR = programId -> {
+    // no-op
+  };
+  private final SystemProgramManagementService systemProgramManagementService;
+  private final ApplicationLifecycleService applicationLifecycleService;
+  private final ProgramLifecycleService programLifecycleService;
+  private final NamespaceAdmin namespaceAdmin;
+  private final CapabilityReader capabilityReader;
+  private final CapabilityWriter capabilityWriter;
+  private final MetadataSearchClient metadataSearchClient;
+
+  @Inject
+  CapabilityApplier(CConfiguration cConf, SystemProgramManagementService systemProgramManagementService,
+                    ApplicationLifecycleService applicationLifecycleService, NamespaceAdmin namespaceAdmin,
+                    ProgramLifecycleService programLifecycleService, CapabilityReader capabilityReader,
```

Comment:
I think this class should take the `CapabilityStatusStore` instead of separate reader and writer.

Reply:
Done